pkg/appclient/jsonrpc.go

```diff
@@ -0,0 +1,116 @@
+	sharedtypes "pocket/x/shared/types"
+)
+
+// handleJSONRPCRelays handles JSON RPC relay requests.
+func (app *appClient) handleJSONRPCRelays(
+	ctx context.Context,
+	serviceId string,
+	request *http.Request,
+) (responseBz []byte, err error) {
+	// Read the request body bytes.
+	payloadBz, err := io.ReadAll(request.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	// Hash and sign the request payload.
+	hash := crypto.Sha256(payloadBz)
+	signature, _, err := app.keyring.Sign(app.keyName, hash)
+	if err != nil {
+		return nil, err
+	}
+
+	// Create the relay request payload.
+	relayRequestPayload := &types.RelayRequest_JsonRpcPayload{}
+	relayRequestPayload.JsonRpcPayload.Unmarshal(payloadBz)
+
+	// Get the current block height to query for the current session.
+	currentBlock := app.blockClient.LatestBlock(ctx)
+
+	// Query for the current session.
+	sessionQueryReq := sessiontypes.QueryGetSessionRequest{
+		ApplicationAddress: app.appAddress,
+		ServiceId:          &sharedtypes.ServiceId{Id: serviceId},
+		BlockHeight:        currentBlock.Height(),
+	}
+	sessionQueryRes, err := app.sessionQuerier.GetSession(ctx, &sessionQueryReq)
+	if err != nil {
+		return nil, err
+	}
+
+	session := sessionQueryRes.Session
+
+	// Get a supplier URL and address for the given service and session.
+	supplierUrl, supplierAddress, err := app.getRelayerUrl(ctx, serviceId, session)
+	if err != nil {
+		return nil, err
+	}
+
+	// Create the relay request.
+	relayRequest := &types.RelayRequest{
+		Meta: &types.RelayRequestMetadata{
+			SessionHeader: session.Header,
+			Signature:     signature,
+		},
+		Payload: relayRequestPayload,
+	}
+
+	// Marshal the relay request to bytes and create a reader to be used as an HTTP request body.
+	relayRequestBz, err := relayRequest.Marshal()
+	if err != nil {
+		return nil, err
+	}
+	relayRequestReader := io.NopCloser(bytes.NewReader(relayRequestBz))
+
+	// Create the HTTP request to send the request to the relayer.
+	relayHTTPRequest := &http.Request{
+		Method: request.Method,
+		Header: request.Header,
+		URL:    supplierUrl,
+		Body:   relayRequestReader,
+	}
+
+	// Perform the HTTP request to the relayer.
+	relayHTTPResponse, err := http.DefaultClient.Do(relayHTTPRequest)
+	if err != nil {
+		return nil, err
+	}
+
+	// Read the response body bytes.
+	relayResponseBz, err := io.ReadAll(relayHTTPResponse.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	// Unmarshal the response bytes into a RelayResponse.
+	relayResponse := &types.RelayResponse{}
+	if err := relayResponse.Unmarshal(relayResponseBz); err != nil {
+		return nil, err
+	}
+
+	// Verify the response signature. We use the supplier address that we got from
+	// the getRelayerUrl function since this is the address we are expecting to sign the response.
+	// TODO_TECHDEBT: if the RelayResponse is an internal error response, we should not verify the signature
+	// as in some relayer early failures, it may not be signed by the supplier.
```

Comment:
Wdyt of making `RelayResponse` have a function (where it is the receiver) that verifies the signature?

Reply:
Spoke offline: I think having a `Signer` interface that is used by both the app/gateway client and the supplier with seperate implementations would be better in conjunction with the `GetSignableBytes()` method on the relay parts

Reply:
I'd bias towards a single implementino (or collocated) but either use an "overloading" like technique or params to generalize. 

If it's not too much effort, wold love to see it in an independent PR.

Reply:
Now the signer interface is in I think it will be easier to abstract it away and replace with the reciever method you think of. Will check it out tmrw